### ISSUE TITLE
Refactor the backbone of operator from controller-runtime based to knative/pkg based

### ISSUE THREAD

**Issue by houshengbo** [tektoncd/operator#13]
# Expected Behavior
As we have a long discussion in the issue https://github.com/tektoncd/operator/issues/1, the benefits to have a consistent architecture among all tekton project will pay back in future, in terms of shifting our efforts from one tekton project to another.

A similar project called serving-operator is undertaking the same transition now. I can move on with this transition, if there is no strong objection. @vdemeester @nikhil-thomas @vincent-pli @hrishin @bobcatfish 

# Actual Behavior
We use a controller-runtime based structure generated by operator-sdk.

# Steps to Reproduce the Problem

1. Check the structure of source code.

# Additional Info

**Comment by sthaha** [comment]
Could you please add an acceptance criteria that `operator-sdk` tooling can still be used?

-   [ ] `operator-sdk` tooling can still be used after the refactoring is done to allow for generation of CRD, the zz_deepcopy.go 
- [ ] operator sdk can still be used for running `e2e` tests
- [ ] operator sdk can be used for generating CSV 

@nikhil-thomas please add to this if I have missed any. 

@houshengbo , it will be great if you could present this change in the Work group meeting so that we understand the workflow of releasing operator to `operatorhub.io` once the refactor is done if the refactor renders the `operator-sdk` tooling unusable.

**Comment by houshengbo** [comment]
@sthaha We will eventually get rid of the dependency on operator-sdk. CRD can be reused and generated by any tool you like. operator-sdk will not be a must for operator.
Regarding the generation of other code, we will use script in k8s.io/code-generator to generate deepcopy,client,informer,lister, and use script under knative.dev/pkg to generate injection. All the generated code is put under client folder, like https://github.com/tektoncd/pipeline/tree/master/pkg/client.

Releasing operator to operatorhub.io seems to be another separate process I also need to investigate, but I can present the necessary changes for this transition, comparing to the current source code.

**Comment by sthaha** [comment]
> @sthaha We will eventually get rid of the dependency on operator-sdk. CRD can be reused and generated by any tool you like. operator-sdk will not be a must for this operator.

I am fine with changing/standarising on tools we all use as along as it doesn't make our life any more harder than needed when releasing to operatorhub.io 

> Releasing operator to operatorhub.io seems to be another separate process I also need to investigate, but I can present the necessary changes for this transition, comparing to the current source code.

Seem to me like the biggest pain/hurdle both @nikhil-thomas and I came across (based on our experience publish downstream openshift-operator), so could this refactor wait until they have figured out the step involved to publish to operatorhub.io?

**Comment by vincent-pli** [comment]
Whatever the decision is, discussion and implementation will cost much time (I saw same thing in `knative serving operator` is still not complete), I think we need make this project running first, then do the enhancement, some stuff is still not ready, such as README ... 😄

**Comment by houshengbo** [comment]
@vincent-pli that's why I opened other issues, like integration tests setup, so that we need to safeguard the functionalities of operator.

**Comment by houshengbo** [comment]
@sthaha Is the publish process different based on what framework we use for the operator project? To me, it sounds separate and independent process. Correct?

**Comment by sthaha** [comment]
@houshengbo, you are right, publishing to operatorhub.io is a simple PR with the `CSV` see: https://github.com/operator-framework/community-operators/tree/master/community-operators/openshift-pipelines-operator 

But getting the PR to pass isn't easy .. especially with the operator scorecard getting on your way. 

So, where operator-sdk really helps is with 
- generation of the crd (pretty sure it uses kubebuilder) 
- generation of CSV : see `operator-sdk olm-catalog` 
- exercise the olm workflow (install the operator through the web ui)  - see `Operator Source`https://github.com/openshift/tektoncd-pipeline-operator/blob/master/deploy/install/operator-source.yaml ;  along with `operator-courier` tool

**Comment by sthaha** [comment]
> I think we need make this project running first, then do the enhancement, some stuff is still not ready, such as README ... 😄

Patches welcome :D

**Comment by nikhil-thomas** [comment]
@houshengbo @vdemeester @sthaha 
> the benefits to have a consistent architecture among all tekton project will pay back in future, in terms of shifting our efforts from one tekton project to another.

I see merit in this statement. However, we should not forget what exactly we are doing here. We are embedding repetitive human-operator logic into a controller (not building the controller). Operator-sdk takes care of all the plumbing and gives us a well-defined space to plugin in our custom reconcile logic. From my perspective it is convenient and the development is fast. The learning curve of operator-sdk is shallow. So getting into operator-sdk based operaotor  will be easy for anyone who is working on any other part of tekton.

`Problems once solved, disappears from view.`
That is, we don't know about challenges implicitly taken care of by operator-sdk. Operator-sdk has an active community contributing to it. The operator-framework ecosystem has other pieces like Operator-Lifecycle-Manager, Operator-Courier etc. They might come up with more tooling in the future.

May be the low entry threshold of operator-sdk can be the ramp through which we get new contributors into Tekton.

My humble opinion is, this proposal definitely has its significance. However, we should also think about how much value it is going to provide to the user. I would prioritize user facing features/improvements first.

**Comment by houshengbo** [comment]
**Why is the operator put under Tekton community?**
It is a Tekton project. If not, we can launch the operator effort anywhere else except Tekton, with having to consider any nature of being a Tekton project.

**How to release the operator?**
Shouldn't we release the image, the source code, or any binaries the same way as Tekton pipeline, or any other project? If not, we do not need to launch the operator effort under Tekton. 
In addition, does operatorhub have hurdle stopping operator with non-operator-sdk project being published? Again, I prefer to smooth out the release process for operator by applying the same way as we release tekton pipeline, rather other releasing in operatorhub. Even if we publish in operatorhub, conforming to the tekton releasing process should have priority.

With that being said, tekton operator should enable ko support, and have VALID integration tests to run in the prow job(with the command like `go test -race -v  -count=1 -tags=e2e -timeout=20m ./test/e2e`, one scenario I can see is to create the Config cluster under a test namepsace, verify if pipeline pods are launched, && delete this Config cluster, to see if pipeline pods are removed.) These work items should be done ASAP.

**Comment by houshengbo** [comment]
**Differences between controller-runtime based controller and knative/pkg based controller?**
**Code generation:**
controller-runtime: both operator-sdk and kubebuilder generate code of controller-runtime based controller, api, crds, deepcopy, etc.
knative/pkg: api, crds can be reused, or generated by any tool. Clients are generated by k8s.io/code-generator including deepcopy,client,informer,lister and Injection is created by knative/pkg itself.

**Comment by vdemeester** [comment]
> Shouldn't we release the image, the source code, or any binaries the same way as Tekton pipeline, or any other project? If not, we do not need to launch the operator effort under Tekton.

I do not fully agree with that. Let's question ourselves : what is the way we release the tektoncd/pipeline project the same as other tektoncd project ? The answer is **no**. The cli doesn't need to support `ko` at all as it's only client-side. 

More importantly, being under the tektoncd organization doesn't mean *all* the repository need to share the same tooling for things. It's a *nice to have* thing, but it is definitely not required. We can discuss the better way forward, but we shouldn't keep ourselve to release a first version of the operator because we are currently refactoring the whole codebase. It doesn't make sense to me.

> In addition, does operatorhub have hurdle stopping operator with non-operator-sdk project being published? Again, I prefer to smooth out the release process for operator by applying the same way as we release tekton pipeline, rather other releasing in operatorhub. Even if we publish in operatorhub, conforming to the tekton releasing process should have priority.

As written above, I disagree. Tooling, helper, framework, they come and goes. And as written abvoe, there is no requirement for a `tektoncd` to adopt any specific tooling.

> With that being said, tekton operator should enable ko support, and have VALID integration tests to run in the prow job(with the command like `go test -race -v -count=1 -tags=e2e -timeout=20m ./test/e2e`.

Same as above, that's a *nice to have*. We already have a way to run *valid* end-to-end tests — maybe they are not well-written or fully testing the operator ; but if that's the case we need to focus on having a very good end-to-end coverage before changing/refactoring the operator code. Once we are satisfied with our coverage (whatever tooling we use to run thoses), we can think of refactoring the core of the operator.

> one scenario I can see is to create the Config cluster under a test namepsace, verify if pipeline pods are launched, && delete this Config cluster, to see if pipeline pods are removed.) These work items should be done ASAP.

I agree on the ASAP, as let's have this scenario (and more) in the current e2e tests. Then we can think and discuss of refactoring.

Copying a bit what @nikhil-thomas said : My humble opinion is, this proposal definitely has its significance. However, we should also think about how much value it is going to provide to the user **and** contributors. I would prioritize user facing features/improvements and having a high level of confidence in our tests first.

**Comment by houshengbo** [comment]
Integration tests should definitely be addressed as soon as possible. Does everyone agree on this first??

I really hope we can enable **ko support**(because we can build and install operator in one single command, without worrying about image name and changing the image path), **without interrupting the operator-sdk build process**. In Prow job of integration test, we can benefit from ko. Is that also something everyone can accept?

Regarding the refactoring, it is not a work item we can finish in a fast pace(even if we want to), so the steps can be first making operator-sdk NOT the ONLY framework we support. Is that something folks feel OK with?

**Comment by nikhil-thomas** [comment]
> Integration tests should definitely be addressed as soon as possible. Does everyone agree on this first??

I repeat we should first focus on making the operator more useful first. We can definitely add integrations tests. I don't understand how operator-sdk is stopping us from that.

> I really hope we can enable ko support(because we can build and install operator in one single command

ko is useful, `operator-sdk up local` command is equally useful. 
It runs the controller out-of-cluster and aids with fast and efficient development. When it comes to tests, (operator-sdk) lets run tests with or without images.

> so the steps can be first making operator-sdk NOT the ONLY framework we support. Is that something folks feel OK with?

I don't understand this. From my experience operator-sdk is stays out of the way of a developer writing operational logic. From my perspective, we are using operator-sdk, so that we can focus on the automating operational logic (install, patch, update, repair) of Tekton-Pipelines. The operator has only very basic features at present. We need to focus on collecting pain-points from Tekton users and embedding solutions in the operator.

Remember we are in the same team :) We need to do our best to improve the Tekton user-experience as a CICD tool. So that some one trying out Tekton has an effective starting point.

**Comment by houshengbo** [comment]
I did not say operator-sdk stop people from doing anything.

If we have the `pain points` as identified, open issue or issues. Or ask more people/users to join the community and try out the operator.

What I propose to add integration tests, enable ko support, etc, so far have nothing to do about any major change. It just enable another way to build and test and operator.

If we think it is good to have, that's fine, coz this is something I can help with. It does not stop anything from tackling the pain points(open new issues or ask more people to join). 

If we think it is garbage to have, that's fine, but setup the integration cases, making sure it covers all the current scenarios, so that we do not blind merge anything.

**Comment by houshengbo** [comment]
@vincent-pli @sthaha @nikhil-thomas @vdemeester 
Knative serving operator has accomplished the structural transition from controller-runtime based model to knative/pkg based model: https://github.com/knative/serving-operator/pull/146.

**Comment by nikhil-thomas** [comment]
Great :+1: 
Did they publish an operator to operatorhub.io after the merge

**Comment by houshengbo** [comment]
That has been opened as an issue to be resolved.
